File: android/src/main/java/org/bnb/dw/QuestionActivity.java
package org.bnb.dw;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.widget.Button;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;
import android.widget.Toast;

import org.bnb.dw.core.Choice;
import org.bnb.dw.core.Question;
import org.bnb.dw.core.Quiz;

public class QuestionActivity extends AppCompatActivity {

  private static final String LOG_TAG = "QuestionActivity";
  private Quiz quiz;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_question);
    initQuiz();
    prepareView();
  }

  @Override
  protected void onNewIntent(Intent intent) {
    super.onNewIntent(intent);
    prepareView();
  }

  private void initQuiz() {
    quiz = new Quiz();
  }

  private void prepareView() {
    Question question = quiz.fetchQuestion();
    printQuestion(question);
    prepareChoices(question);
    prepareCheckButton(question);
  }

  private void printQuestion(Question question) {
    TextView text = findViewById(R.id.questionText);
    text.setText(question.text());
  }

  private void prepareChoices(Question question) {
    RadioGroup radioGroup = findViewById(R.id.answersRadioGroup);
    radioGroup.removeAllViews();
    for (Choice choice : question.choices) {
      RadioButton newRadioButton = new RadioButton(this);
      newRadioButton.setText(choice.text);
      newRadioButton.setTag(choice);
      radioGroup.addView(newRadioButton);
    }
  }

  private void prepareCheckButton(Question question) {
    Button checkButton = findViewById(R.id.checkButton);
    checkButton.setOnClickListener(v -> check(question));
    checkButton.setText(R.string.check);
  }

  private void check(Question question) {
    RadioButton selectedChoice = getSelectedChoice();
    if (selectedChoice != null) {
      boolean result = quiz.verify(question, (Choice) selectedChoice.getTag());
      Log.d(LOG_TAG, "Correct? " + result);
      informUser(result);
      if (result)
        enableNavigationFurther();
    }
  }

  private RadioButton getSelectedChoice() {
    RadioGroup radioGroup = findViewById(R.id.answersRadioGroup);
    int checkedRadioButtonId = radioGroup.getCheckedRadioButtonId();
    return checkedRadioButtonId != -1 ? findViewById(checkedRadioButtonId) : null;
  }

  private void informUser(boolean result) {
    CharSequence text;
    if (result) {
      text = "OK";
    } else {
      text = "Errr....";
    }
    Toast toast = Toast.makeText(this, text, Toast.LENGTH_SHORT);
    toast.show();
  }

  private void enableNavigationFurther() {
    Button checkButton = findViewById(R.id.checkButton);
    checkButton.setText(R.string.next);
    checkButton.setOnClickListener(v -> nextQuestion());
  }

  private void nextQuestion() {
    Intent intent = new Intent(this, QuestionActivity.class);
    startActivity(intent);
  }
}

File: android/src/main/kotlin/org/bnb/dw/QuestionActivity.kt
package org.bnb.dw

import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.os.Environment
import android.support.v7.app.AppCompatActivity
import android.support.v7.preference.PreferenceManager
import android.util.Log
import android.widget.RadioButton
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_question.*
import org.bnb.dw.core.Choice
import org.bnb.dw.core.Question
import org.bnb.dw.core.Quiz
import org.bnb.dw.core.Settings

class QuestionActivity : AppCompatActivity() {
    private var quiz: Quiz? = null
    private val repo = FilesRepository(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS).absolutePath + "/dw/")

    private val selectedChoice: RadioButton?
        get() {
            val checkedRadioButtonId = answersRadioGroup.checkedRadioButtonId
            return if (checkedRadioButtonId != -1) findViewById(checkedRadioButtonId) else null
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_question)
        initQuiz()
    }

    override fun onStart() {
        super.onStart()
        prepareView()
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        prepareView()
    }

    private fun initQuiz() {
        if (Environment.getExternalStorageState() in
                setOf(Environment.MEDIA_MOUNTED, Environment.MEDIA_MOUNTED_READ_ONLY)) {
            quiz = Quiz(repo, translateSettings(PreferenceManager.getDefaultSharedPreferences(this)))
        } else {
            Log.e(this.localClassName, "Unable to load data, media not ready")
            throw IllegalStateException("Unable to read media directory")
        }
    }

    private fun translateSettings(preferences: SharedPreferences): Settings {
        val settings = Settings()
        if (preferences.contains(getString(R.string.pref_nouns_gender_weight)))
            settings.genderWeight = preferences.getInt(getString(R.string.pref_nouns_gender_weight), -1)
        if (preferences.contains(getString(R.string.pref_nouns_noun_weight)))
            settings.nounWeight = preferences.getInt(getString(R.string.pref_nouns_noun_weight), -1)
        if (preferences.contains(getString(R.string.pref_nouns_translation_weight)))
            settings.translationWeight = preferences.getInt(getString(R.string.pref_nouns_translation_weight), -1)
        return settings
    }

    private fun prepareView() {
        val question = quiz!!.fetchQuestion()
        printQuestion(question)
        prepareChoices(question)
        prepareCheckButton(question)
    }

    private fun printQuestion(question: Question) {
        questionText.text = question.text()
    }

    private fun prepareChoices(question: Question) {
        answersRadioGroup.removeAllViews()
        for (choice in question.choices) {
            val newRadioButton = RadioButton(this)
            newRadioButton.text = choice.text
            newRadioButton.tag = choice
            answersRadioGroup.addView(newRadioButton)
        }
    }

    private fun prepareCheckButton(question: Question) {
        checkButton.setOnClickListener { check(question) }
        checkButton.setText(R.string.check)
    }

    private fun check(question: Question) {
        val selectedChoice = selectedChoice
        if (selectedChoice != null) {
            val result = quiz!!.verify(question, selectedChoice.tag as Choice)
            informUser(result)
            if (result)
                nextQuestion()
        }
    }

    private fun informUser(result: Boolean) {
        val text = if (result) "OK" else "Errr...."
        val toast = Toast.makeText(this, text, Toast.LENGTH_SHORT)
        toast.show()
    }

    private fun nextQuestion() {
        val intent = Intent(this, QuestionActivity::class.java)
        startActivity(intent)
    }
}
